src/platform/datapath_linux.c

```diff
@@ -1364,51 +1381,53 @@ CxPlatSocketContextPendSend(
             return errno;
         }
 
-        if (LocalAddress != NULL) {
-            CxPlatCopyMemory(
-                &SendContext->LocalAddress,
-                LocalAddress,
-                sizeof(*LocalAddress));
-            SendContext->Bind = TRUE;
-        }
-
-        CxPlatCopyMemory(
-            &SendContext->RemoteAddress,
-            RemoteAddress,
-            sizeof(*RemoteAddress));
-
         SocketContext->SendWaiting = TRUE;
     }
 
     if (SendContext->Pending) {
         //
-        // This was a send that was already pending, so we need to add it back
-        // to the head of the queue.
+        // A pending send is not removed from the list before attempting to send it.
         //
-        CxPlatListInsertHead(
-            &SocketContext->PendingSendContextHead,
-            &SendContext->PendingSendLinkage);
-    } else {
-        //
-        // This is a new send that wasn't previously pended. Add it to the end
-        // of the queue.
-        //
-        CxPlatListInsertTail(
-            &SocketContext->PendingSendContextHead,
-            &SendContext->PendingSendLinkage);
-        SendContext->Pending = TRUE;
+        goto Exit;
     }
 
+    if (LocalAddress != NULL) {
+        CxPlatCopyMemory(
+            &SendContext->LocalAddress,
+            LocalAddress,
+            sizeof(*LocalAddress));
+        SendContext->Bind = TRUE;
+    }
+
+    CxPlatCopyMemory(
+        &SendContext->RemoteAddress,
+        RemoteAddress,
+        sizeof(*RemoteAddress));
+
+    //
+    // This is a new send that wasn't previously pended. Add it to the end
+    // of the queue.
+    //
+    CxPlatListInsertTail(
+        &SocketContext->PendingSendContextHead,
+        &SendContext->PendingSendLinkage);
+    SendContext->Pending = TRUE;
+
+Exit:
     return QUIC_STATUS_SUCCESS;
 }
 
+//
+// N.B. Caller must be holding SocketContext->PendingSendContextLock Lock.
+//
 QUIC_STATUS
 CxPlatSocketContextSendComplete(
```

Comment:
Can we get a design where this doesn't need the lock to be held, and instead it only grabs where necessary? I don't like this whole thing requiring the lock; and especially don't like where it selectively unlocks as necessary.

Reply:
There is going to be no way to avoid the unlock and lock in the middle of this. We need to do the epoll part under the lock, in case it races with the worker thread. I can look a bit more into making the caller not hold the lock though.

Reply:
At the very least, I don't want the caller to grab the lock. There is no need. Let this function handle it. On the topic of epoll, I don't see why the lock needs to be held for that. Generally, you should never be required to hold any lock and call into some other layer/component.

Reply:
Done.